pkg/controller/clusterprovision/clusterprovision_controller.go

```diff
@@ -0,0 +1,319 @@
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("clusterprovision-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to ClusterProvision
+	err = c.Watch(&source.Kind{Type: &hivev1.ClusterProvision{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to ClusterDeployment
+	err = c.Watch(&source.Kind{Type: &hivev1.ClusterDeployment{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(
+			func(a handler.MapObject) []reconcile.Request {
+				return selectorDeploymentWatchHandler(a, mgr)
+			}),
+	})
+
+	// Watch for install jobs created for ClusterProvision
+	err = c.Watch(&source.Kind{Type: &batchv1.Job{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &hivev1.ClusterProvision{},
+	})
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileClusterProvision{}
+
+// ReconcileClusterProvision reconciles a ClusterProvision object
+type ReconcileClusterProvision struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a ClusterProvision object and makes changes based on the state read
+// and what is in the ClusterProvision.Spec
+// +kubebuilder:rbac:groups=hive.openshift.io,resources=ClusterProvisions,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=hive.openshift.io,resources=ClusterProvisions/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=hive.openshift.io,resources=ClusterDeployment,verbs=get;list;watch
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=configmaps;namespaces,verbs=get
+func (r *ReconcileClusterProvision) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	pLog := log.WithFields(log.Fields{
+		"name":       request.NamespacedName.String(),
+		"controller": controllerName,
+	})
+	// For logging, we need to see when the reconciliation loop starts and ends.
+	pLog.Info("reconciling cluster provision")
+	// Fetch the ClusterProvision instance
+	instance := &hivev1.ClusterProvision{}
+	err := r.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			pLog.Debug("ClusterProvision not found, skipping")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		pLog.WithError(err).Error("cannot get ClusterProvision")
+		return reconcile.Result{}, err
+	}
+
+	if !instance.DeletionTimestamp.IsZero() {
+		pLog.Debug("ClusterProvision being deleted, skipping")
+		return reconcile.Result{}, nil
+	}
+
+	switch instance.Spec.Stage {
+	case hivev1.ClusterProvisionStageComplete, hivev1.ClusterProvisionStageFailed:
+		return r.deleteJob(instance, pLog)
+	}
+
+	deployment := &hivev1.ClusterDeployment{}
```

Comment:
deployment a little unclear here, I'd go with out standard and very concise "cd" var name.